src/clients/FtpGateway.cc

```diff
@@ -1008,6 +1008,11 @@ Ftp::Gateway::processReplyBody()
 
     entry->flush();
 
+    if (theSize >= 0 && data.payloadSeen >= theSize) {
+        markParsedVirginReplyAsWhole("whole virgin body");
+        completeForwarding();
```

Comment:
Old theSize and new data.payloadSeen measure virgin bytes. StoreEntry stores adapted bytes. We should not completeForwarding() if we are still adapting received virgin bytes, even if we have received all of those virgin bytes.

Secondary red flags that may go away when the primary concern above is addressed:

* Is theSize guaranteed to be known for all relevant cases?
* This code does not cover flags.isdir path, and we are not adding any other code to cover that path. Perhaps that path calls completeForwarding() correctly already but it is a red flag.